app/services/control_rate_service.rb

```diff
@@ -67,19 +75,26 @@ def call
   end
 
   def sum_cumulative_registrations
-    initial_count = region.registered_patients.with_hypertension.where("recorded_at < ?", periods.begin.to_date).count
-    periods.each_with_object({}) { |period, running_totals|
-      previous_registrations = running_totals[period.previous] || initial_count
+    earliest_registration_period = [periods.begin, registration_counts.keys.first].compact.min
+    (earliest_registration_period..periods.end).each_with_object({}) { |period, running_totals|
+      previous_registrations = running_totals[period.previous] || 0
       current_registrations = registration_counts[period] || 0
-      running_totals[period] = current_registrations + previous_registrations
+      total = current_registrations + previous_registrations
+      running_totals[period] = total
     }
   end
 
   def registration_counts
-    range = (periods.begin.start_date..periods.end.end_date)
+    return @registration_counts if @registration_counts
     formatter = lambda { |v| quarterly_report? ? Period.quarter(v) : Period.month(v) }
-    result = region.registered_patients.with_hypertension.group_by_period(periods.begin.type, :recorded_at, {range: range, format: formatter}).count
-    result.drop_while { |period, count| count == 0 }.to_h
+    result = region.registered_patients.with_hypertension.group_by_period(periods.begin.type, :recorded_at, {format: formatter}).count
```

Comment:
We now grab _all_ registrations w/ the same query, so we don't have to worry about adding in the registrations that happen before the `periods` range.

Reply:
I like this a lot. Grab everything and chop off what wasn't requested. Let groupdate take care of all the annoying boundary conditions and timezones and stuff.